packages/contracts/test/withdrawer.spec.ts

```diff
@@ -0,0 +1,213 @@
+          sender: await l2Signer.getAddress(),
+          target: NON_ZERO_ADDRESS,
+          value: 0,
+          gasLimit: NON_ZERO_GASLIMIT,
+          data: NON_ZERO_DATA,
+        })
+      )
+    })
+
+    it('Should add an entry to the withdrawals mapping', async () => {
+      const nonceAfter = await withdrawer.nonce()
+      expect(await withdrawer.withdrawals(withdrawalHash)).to.be.true
+      expect(nonceAfter.sub(nonceBefore).toNumber()).to.eq(1)
+    })
+
+    // Test to ensure we're correctly calculating the storageKey. Per the solidity docs:
+    // "The value corresponding to a mapping key k is located at keccak256(h(k) . p) where . is
+    //   concatenation and h is a function that is applied to the key..."
+    it('Should return bytes32(1) when querying the calculated storage key', async () => {
+      const storageSlot = '00'.repeat(31) + '01' // i.e the second variable declared in the contract
+      storageKey = ethers.utils.keccak256(withdrawalHash + storageSlot)
+
+      expect(
+        await l2GethProvider.getStorageAt(withdrawerAddress, storageKey)
+      ).to.equal(utils.hexZeroPad('0x01', 32))
+    })
+
+    it('should generate a valid proof', async () => {
+      // Get the proof
+      proof = await l2GethProvider.send('eth_getProof', [
+        withdrawerAddress,
+        [storageKey],
+        toRpcHexString((await l2GethProvider.getBlock('latest')).number),
+      ])
+
+      // Sanity check expected values of the proof
+      expect(proof.storageProof[0].key).to.eq(storageKey)
+      expect(proof.storageProof[0].value).to.eq('0x1')
+
+      // Check the proof directly against the SecureMerkleTrie lib
+      expect(
+        await testLibSecureMerkleTrie.verifyInclusionProof(
+          proof.storageProof[0].key,
+          '0x01',
+          toHexString(rlp.encode(proof.storageProof[0].proof)),
+          proof.storageHash
+        )
+      ).to.be.true
+    })
+  })
+
+  describe('Finalizing a withdrawal on L1', () => {
+    let withdrawalVerifier: WithdrawalVerifier
+    let l2OutputOracle: L2OutputOracle
+    let l2Timestamp: BigNumber
+    before('Setup L1 system contracts', async () => {
+      // Deploy the L2 Output Oracle with a 1 second submission interval so that we can immediately
+      // test the withdrawal.
+      l2OutputOracle = await (
+        await new L2OutputOracle__factory(l1Signer)
+      ).deploy(
+        1, // submissionInterval
+        1, // l2BlockTime
+        ethers.utils.keccak256('0x00'), // genesisL2Output
+        100, // historicalTotalBlocks
+        signerAddress
+      )
+
+      // Deploy the WithdrawalVerifier with a 0 second finalization delay.
+      withdrawalVerifier = await (
+        await new WithdrawalVerifier__factory(l1Signer)
+      ).deploy(l2OutputOracle.address, 0)
+
+      // create an output root that we can prove against only the storage root matters
+      // for our purposes.
+      const outputRoot = ethers.utils.keccak256(
```

Comment:
This kind of thing should be turned into a helper function. There should be a ts lib that has the same methods as the solidity libs used